### PR TITLE
Ed25519verify

### PR DESCRIPTION
**Pull request by stechu**
add a new op code `ed25519verify` which checks a signature using ed25519 crypto signature scheme.

benchmark result:

```
goos: darwin
goarch: amd64
pkg: github.com/algorand/go-algorand/data/transactions/logic
BenchmarkEd25519Verify-8   	   30000	     57359 ns/op
PASS
ok  	github.com/algorand/go-algorand/data/transactions/logic	3.932s
```

### CODE REVIEW

File: data/transactions/logic/eval.go [hunk 1]
```diff
@@ -24,6 +24,7 @@ import (
 	"encoding/hex"
 	"errors"
 	"fmt"
+	"github.com/algorand/go-algorand/crypto"
```
Comment (algobolson) [inline data/transactions/logic/eval.go]: group this with the other similar imports

File: data/transactions/logic/eval.go [hunk 2]
```diff
@@ -885,3 +887,32 @@ func opGlobal(cx *evalContext) {
...
 }
+
+func opEd25519verify(cx *evalContext){
+	last := len(cx.stack) - 1 // index of PK
+	prev := last -1           // index of signature
+	pprev := prev - 1         // index of data
+
+	var sv crypto.SignatureVerifier
+	if len(cx.stack[last].Bytes) != len(sv) {
+		cx.err = fmt.Errorf("invalid public key")
```
Comment (algobolson) [inline data/transactions/logic/eval.go]: errors.New("") for constant strings that don't need formatting

File: data/transactions/logic/eval.go [hunk 3]
```diff
@@ -885,3 +887,32 @@ func opGlobal(cx *evalContext) {
...
+
+	var sig crypto.Signature
+	if len(cx.stack[prev].Bytes) != len(sig) {
+		cx.err = fmt.Errorf("invalid signature")
+		return
+	}
+	copy(sig[:], cx.stack[prev].Bytes)
+
+	if sv.VerifyBytes(cx.stack[pprev].Bytes, sig) {
+		cx.stack[pprev].Bytes = nil
```
Comment (algobolson) [inline data/transactions/logic/eval.go]: move common line outside of if/else

File: data/transactions/logic/eval_test.go [hunk 4]
```diff
@@ -20,6 +20,8 @@ import (
 	"encoding/base64"
 	"encoding/binary"
 	"encoding/hex"
+	"fmt"
+	"github.com/algorand/go-algorand/crypto"
```
Comment (algobolson) [inline data/transactions/logic/eval_test.go]: group this with similar imports

File: data/transactions/logic/eval_test.go [hunk 5]
```diff
@@ -1281,3 +1283,77 @@ func BenchmarkSha512_256(b *testing.B) {
...
+	pass1, err := Eval(program, ep1)
+	require.False(t, pass1)
+	require.NoError(t, err)
+}
+
+func BenchmarkEd25519Verify(b *testing.B){
+	sb := strings.Builder{}
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		b.StopTimer()
```
Comment (algobolson) [inline data/transactions/logic/eval_test.go]: move setup to a separate loop. it should be okay to allocate b.N test [32]byte data values and [64]byte sigs. then run the benchmark in a loop with the timer on for the whole set.

**Review by algobolson — CHANGES_REQUESTED** [review]

File: data/transactions/logic/eval.go (hunk 1 above)
Comment (stechu) [reply data/transactions/logic/eval.go]: fixed.

File: data/transactions/logic/eval.go (hunk 2 above)
Comment (stechu) [reply data/transactions/logic/eval.go]: fixed.

File: data/transactions/logic/eval.go (hunk 3 above)
Comment (stechu) [reply data/transactions/logic/eval.go]: fixed.

File: data/transactions/logic/eval_test.go (hunk 4 above)
Comment (stechu) [reply data/transactions/logic/eval_test.go]: fixed.

File: data/transactions/logic/eval_test.go (hunk 5 above)
Comment (stechu) [reply data/transactions/logic/eval_test.go]: done. interestingly, the time is a bit faster.

File: data/transactions/logic/eval.go [hunk 6]
```diff
@@ -24,11 +24,11 @@ import (
 	"encoding/hex"
 	"errors"
 	"fmt"
-	"io"
-
 	"golang.org/x/crypto/sha3"
```
Comment (algobolson) [inline data/transactions/logic/eval.go]: keep the gap. there should be 3 groups. standard libraries, other libraries, our code.

**Review by algobolson — CHANGES_REQUESTED** [review]

**Pull request merged**